lib/pairmotron/pair_maker.ex

```diff
@@ -58,23 +69,50 @@ defmodule Pairmotron.PairMaker do
       |> Mixer.mixify(week)
       |> Pairer.generate_pairs(pairs)
 
-    results.pairs
-      |> Enum.map(fn(users) -> make_pairs(users, year, week, group_id) end)
-      |> List.flatten
-      |> Enum.map(fn(p) -> Repo.insert! p end)
+    Ecto.Multi.new
+      |> remove_dead_pairs(determination.dead_pairs)
+      |> insert_pairs(results.pairs, year, week, group_id)
+      |> insert_lone_user_pair(results.user_pair)
+      |> commit_transaction(year, week, group_id)
+  end
+
+  defp remove_dead_pairs(multi, dead_pairs) do
+    dead_pairs
+      |> Enum.reduce(multi, fn(pair, acc) ->
+          acc |> Ecto.Multi.delete(:remove_dead_pairs, pair)
+      end)
+  end
 
-    insert_user_pair(results.user_pair)
+  defp insert_lone_user_pair(multi, nil), do: multi
+  defp insert_lone_user_pair(multi, user_pair) do
+    multi |> Ecto.Multi.insert(:create_lone_user_pair, user_pair)
   end
 
-  defp insert_user_pair(nil), do: nil
-  defp insert_user_pair(user_pair) do
-    Repo.insert! user_pair
+  defp insert_pairs(multi, pairs, year, week, group_id) do
+    pairs
+      |> Enum.reduce(multi, fn(users, acc) ->
+          insert_pair(acc, year, week, group_id, users)
+      end)
   end
 
-  defp make_pairs(users, year, week, group_id) do
-    pair = Repo.insert! Pair.changeset(%Pair{}, %{year: year, week: week, group_id: group_id})
-    users
-      |> Enum.map(fn(user) -> UserPair.changeset(%UserPair{}, %{pair_id: pair.id, user_id: user.id}) end)
+  defp insert_pair(multi, year, week, group_id, users) do
+    atom = String.to_atom("insert_pair_#{year}_#{week}_#{group_id}_#{hd(users).id}")
```

Comment:
Are we sure it's going to be a good idea to dynamically generate atoms like this? Atom's aren't garbage collected and there is a upper limit to the number of atoms that are allowed to exist (I think it's a million). It'd be pretty crazy to run into that limit, but still.